File: contracts.ts
export interface PlanningHorizon {
    start_datetime: string;
    horizon_minutes: number;
    timezone: string;
}

export interface BlockSection {
    block_section_id: string;
    name: string;
    division: string;
    adjacent_section_ids: string[];
    track_type: string;
}

export interface TrainMovement {
    train_id: string;
    block_section_id: string;
    direction: string;
    scheduled_entry: string;
    scheduled_exit: string;
    headway_before_minutes: number;
    headway_after_minutes: number;
    flexibility: "fixed" | "semi_flexible";
    max_shift_minutes: number;
    train_priority_class: string;
}

export interface PriorityInputs {
    severity_S: number;
    escalation_risk_R: number;
    criticality_C: number;
    age_A: number;
    opportunity_O: number;
}

export interface RouteLeg {
    leg_id: number;
    block_section_id: string;
    duration_minutes: number;
}

export interface MLEnrichment {
    escalation_risk_probability: number;
    escalation_risk_R: number;
    risk_confidence: "high" | "low_model_confidence" | "fallback_rule_based";
    top_risk_factors: string[];
    predicted_duration_minutes: number;
    duration_quantile: string;
    duration_source: "model_p80" | "static_fallback";
    cluster_id?: string;
    model_version: string;
}

export interface MaintenanceRequest {
    work_id: string;
    source_system: string;
    defect_type: string;
    job_type?: "single" | "traveling";
    block_sections_required?: string[];
    duration_minutes?: number;
    route_legs?: RouteLeg[];
    max_wait_minutes?: number;
    setup_minutes?: number;
    earliest_start: string;
    latest_start: string;
    mandatory: boolean;
    crew_type?: string;
    crew_size?: number;
    plant_type?: string;
    plant_qty?: number;
    isolation_group?: string;
    precedence_after?: string[];
    priority_inputs?: PriorityInputs;
    priority_score: number;
    ml_enrichment?: MLEnrichment;
}

export interface CrewShift {
    start: string;
    end: string;
    capacity: number;
}

export interface CrewPool {
    crew_type: string;
    shifts: CrewShift[];
}

export interface PlantPool {
    plant_type: string;
    total_qty: number;
}

export interface ResourceCapacity {
    crew_pools: CrewPool[];
    plant_pools: PlantPool[];
}

export interface SolverConfig {
    max_time_in_seconds?: number;
    num_search_workers?: number;
    random_seed?: number;
    max_number_of_conflicts?: number;
}

export interface DailyCapacityCalendar {
    date: string;
    crew_hours: Record<string, number>;
    plant_hours: Record<string, number>;
    corridor_possession_hours: Record<string, number>;
}

export interface ScheduleRequest {
    planning_horizon: PlanningHorizon;
    block_sections: BlockSection[];
    train_movements: TrainMovement[];
    maintenance_requests: MaintenanceRequest[];
    resource_capacity: ResourceCapacity;
    solver_config?: SolverConfig;
    daily_capacity_calendar?: DailyCapacityCalendar[];
}

export type SolverStatus = 'OPTIMAL' | 'FEASIBLE' | 'INFEASIBLE' | 'FALLBACK_HEURISTIC';

export interface SelectedWorkPackage {
    work_id: string;
    selected: boolean;
    assigned_start: string;
    assigned_end: string;
    block_sections_used: string[];
    crew_assigned?: Record<string, any>;
    plant_assigned?: Record<string, any>;
    bundle_group_id?: string;
    priority_score: number;
}

export interface TravelingJobLegSchedule {
    block_section_id: string;
    start: string;
    end: string;
}

export interface TravelingJobSchedule {
    work_id: string;
    selected: boolean;
    legs: TravelingJobLegSchedule[];
    crew_assigned?: Record<string, any>;
    plant_assigned?: Record<string, any>;
    priority_score: number;
}

export type ReasonCode = 'NO_FEASIBLE_WINDOW' | 'CREW_CAPACITY_EXCEEDED' | 'PLANT_CAPACITY_EXCEEDED' | 'LOWER_PRIORITY' | 'NO_CAPACITY_BEFORE_DEADLINE';

export interface DeferredWorkPackage {
    work_id: string;
    selected: boolean;
    reason_code: ReasonCode;
    reason_detail: string;
    priority_score: number;
    suggested_next_window?: string;
}

export interface GanttTask {
    id: string;
    text: string;
    start_date: string;
    end_date: string;
    parent: string;
    type: string;
    priority_color: string;
    conflict: boolean;
}

export interface Conflict {
    conflict_id: string;
    type: string;
    involved_work_ids: string[];
    block_section_id?: string;
    description: string;
}

export interface KPIs {
    train_delay_minutes_total: number;
    work_packages_included_pct: number;
    unused_block_minutes: number;
    mandatory_items_dropped: number;
}

export interface ScheduleResponse {
    schedule_id: string;
    generated_at: string;
    solver_status: SolverStatus;
    objective_value?: number;
    best_bound?: number;
    optimality_gap_pct?: number;
    solve_time_ms: number;
    selected_work_packages: SelectedWorkPackage[];
    traveling_job_schedules: TravelingJobSchedule[];
    deferred_work_packages: DeferredWorkPackage[];
    gantt_tasks: GanttTask[];
    conflicts: Conflict[];
    kpis: KPIs;
}

export interface DayAllocation {
    date: string;
    assigned_work_ids: string[];
}

export interface UnassignableWork {
    work_id: string;
    reason_code: ReasonCode;
    mandatory: boolean;
}

export interface MacroAllocationCalendar {
    generated_at: string;
    horizon_days: number;
    days: DayAllocation[];
    unassignable_before_deadline: UnassignableWork[];
}

File: backend/app/schemas/contracts.py
from typing import List, Optional, Literal, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum

class PlanningHorizon(BaseModel):
    start_datetime: str
    horizon_minutes: int
    timezone: str

class BlockSection(BaseModel):
    block_section_id: str
    name: str
    division: str
    adjacent_section_ids: List[str]
    track_type: str

class TrainMovement(BaseModel):
    train_id: str
    block_section_id: str
    direction: str
    scheduled_entry: str
    scheduled_exit: str
    headway_before_minutes: int
    headway_after_minutes: int
    flexibility: Literal["fixed", "semi_flexible"]
    max_shift_minutes: int
    train_priority_class: str

class PriorityInputs(BaseModel):
    severity_S: float
    escalation_risk_R: float
    criticality_C: float
    age_A: float
    opportunity_O: float

class RouteLeg(BaseModel):
    leg_id: int
    block_section_id: str
    duration_minutes: int

class MLEnrichment(BaseModel):
    escalation_risk_probability: float
    escalation_risk_R: float
    risk_confidence: Literal["high", "low_model_confidence", "fallback_rule_based"]
    top_risk_factors: List[str]
    predicted_duration_minutes: int
    duration_quantile: str
    duration_source: Literal["model_p80", "static_fallback"]
    cluster_id: Optional[str] = None
    model_version: str

class MaintenanceRequest(BaseModel):
    work_id: str
    source_system: str
    defect_type: str
    job_type: Optional[Literal["single", "traveling"]] = None
    block_sections_required: Optional[List[str]] = None
    duration_minutes: Optional[int] = None
    route_legs: Optional[List[RouteLeg]] = None
    max_wait_minutes: Optional[int] = None
    setup_minutes: Optional[int] = None
    earliest_start: str
    latest_start: str
    mandatory: bool
    crew_type: Optional[str] = None
    crew_size: Optional[int] = None
    plant_type: Optional[str] = None
    plant_qty: Optional[int] = None
    isolation_group: Optional[str] = None
    precedence_after: Optional[List[str]] = Field(default_factory=list)
    priority_inputs: Optional[PriorityInputs] = None
    priority_score: float
    ml_enrichment: Optional[MLEnrichment] = None

class CrewShift(BaseModel):
    start: str
    end: str
    capacity: int

class CrewPool(BaseModel):
    crew_type: str
    shifts: List[CrewShift]

class PlantPool(BaseModel):
    plant_type: str
    total_qty: int

class ResourceCapacity(BaseModel):
    crew_pools: List[CrewPool]
    plant_pools: List[PlantPool]

class SolverConfig(BaseModel):
    max_time_in_seconds: float = 8.0
    num_search_workers: int = 8
    random_seed: int = 42
    max_number_of_conflicts: int = 100000

class DailyCapacityCalendar(BaseModel):
    date: str
    crew_hours: Dict[str, int]
    plant_hours: Dict[str, int]
    corridor_possession_hours: Dict[str, int]

class ScheduleRequest(BaseModel):
    planning_horizon: PlanningHorizon
    block_sections: List[BlockSection]
    train_movements: List[TrainMovement]
    maintenance_requests: List[MaintenanceRequest]
    resource_capacity: ResourceCapacity
    solver_config: SolverConfig = Field(default_factory=SolverConfig)
    daily_capacity_calendar: Optional[List[DailyCapacityCalendar]] = None

class SolverStatus(str, Enum):
    OPTIMAL = "OPTIMAL"
    FEASIBLE = "FEASIBLE"
    INFEASIBLE = "INFEASIBLE"
    FALLBACK_HEURISTIC = "FALLBACK_HEURISTIC"

class SelectedWorkPackage(BaseModel):
    work_id: str
    selected: bool
    assigned_start: str
    assigned_end: str
    block_sections_used: List[str]
    crew_assigned: Optional[Dict[str, Any]] = None
    plant_assigned: Optional[Dict[str, Any]] = None
    bundle_group_id: Optional[str] = None
    priority_score: float

class TravelingJobLegSchedule(BaseModel):
    block_section_id: str
    start: str
    end: str

class TravelingJobSchedule(BaseModel):
    work_id: str
    selected: bool
    legs: List[TravelingJobLegSchedule]
    crew_assigned: Optional[Dict[str, Any]] = None
    plant_assigned: Optional[Dict[str, Any]] = None
    priority_score: float

class ReasonCode(str, Enum):
    NO_FEASIBLE_WINDOW = "NO_FEASIBLE_WINDOW"
    CREW_CAPACITY_EXCEEDED = "CREW_CAPACITY_EXCEEDED"
    PLANT_CAPACITY_EXCEEDED = "PLANT_CAPACITY_EXCEEDED"
    LOWER_PRIORITY = "LOWER_PRIORITY"
    NO_CAPACITY_BEFORE_DEADLINE = "NO_CAPACITY_BEFORE_DEADLINE"

class DeferredWorkPackage(BaseModel):
    work_id: str
    selected: bool
    reason_code: ReasonCode
    reason_detail: str
    priority_score: float
    suggested_next_window: Optional[str] = None

class GanttTask(BaseModel):
    id: str
    text: str
    start_date: str
    end_date: str
    parent: str
    type: str
    priority_color: str
    conflict: bool

class Conflict(BaseModel):
    conflict_id: str
    type: str
    involved_work_ids: List[str]
    block_section_id: Optional[str] = None
    description: str

class KPIs(BaseModel):
    train_delay_minutes_total: int
    work_packages_included_pct: float
    unused_block_minutes: int
    mandatory_items_dropped: int

class ScheduleResponse(BaseModel):
    schedule_id: str
    generated_at: str
    solver_status: SolverStatus
    objective_value: Optional[float] = None
    best_bound: Optional[float] = None
    optimality_gap_pct: Optional[float] = None
    solve_time_ms: int
    selected_work_packages: List[SelectedWorkPackage]
    traveling_job_schedules: List[TravelingJobSchedule]
    deferred_work_packages: List[DeferredWorkPackage]
    gantt_tasks: List[GanttTask]
    conflicts: List[Conflict]
    kpis: KPIs

class DayAllocation(BaseModel):
    date: str
    assigned_work_ids: List[str]

class UnassignableWork(BaseModel):
    work_id: str
    reason_code: ReasonCode
    mandatory: bool

class MacroAllocationCalendar(BaseModel):
    generated_at: str
    horizon_days: int
    days: List[DayAllocation]
    unassignable_before_deadline: List[UnassignableWork]
